### PR TITLE
[SPARK-8685]  [SQL] [PySpark] let Row workwith duplicated column names

### PR DESCRIPTION
**Pull request by davies**

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -1236,6 +1230,7 @@ class Row(tuple):
 
         """ Row in DataFrame """
         __datatype = dataType
```
Comment (rxin) [inline python/pyspark/sql/types.py]: can you add inline comment for all the private fields here on what they do? 

this includes __datatype, __converters__, __fields__, __slots__

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -1160,18 +1150,22 @@ def _has_struct_or_date(dt):
 def _create_properties(fields):
     """Create properties according to fields"""
```
Comment (rxin) [inline python/pyspark/sql/types.py]: we need to expand this comment as well to more clearly define what is happening with this function.

Comment (rxin) [inline python/pyspark/sql/types.py]: basically my worry is that there is a lot of magic happening with the python implementation, and it is very hard to figure out what's going on without extensive comments.

**Comment by davies** [comment]
Close this in favor of #7301

**Pull request closed without merging**